// @volar-plugins/prettyhtml/out/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const prettyhtml_1 = __importDefault(require("@starptech/prettyhtml"));
exports.default = (configs) => () => ({
    provideDocumentFormattingEdits(document, range, options) {
        if (document.languageId !== 'html')
            return;
        const oldRangeText = document.getText(range);
        const newRangeText = (0, prettyhtml_1.default)(oldRangeText, {
            ...configs,
            tabWidth: options.tabSize,
            useTabs: !options.insertSpaces,
        }).contents;
        if (newRangeText === oldRangeText)
            return [];
        const newText = document.getText({
            start: document.positionAt(0),
            end: range.start,
        })
            + newRangeText
            + document.getText({
                start: range.end,
                end: document.positionAt(document.getText().length),
            });
        return [{
                newText,
                range: {
                    start: document.positionAt(0),
                    end: document.positionAt(document.getText().length),
                },
            }];
    },
});
