// @shopify/cli-kit/dist/public/node/vendor/dev_server/network/index.js
import { spawnSync } from 'node:child_process';
export { getIpFromHosts } from './host.js';
// eslint-disable-next-line prettier/prettier
const DEFAULT_CONNECT_TIMEOUT = 1000;
// Skip initialization on module load to prevent Spin trying to load a macOS dylib
// (port checks should never run on Spin anyway)
let checkPort;
export function assertConnectable(options) {
    checkPort || (checkPort = getCheckPortHelper());
    const { port, addr, timeout = DEFAULT_CONNECT_TIMEOUT } = options;
    try {
        const normalizedAddr = addr === 'localhost' ? '127.0.0.1' : addr;
        const running = checkPort(normalizedAddr, port, timeout);
        if (!running) {
            throw new Error(`DevServer for '${options.projectName}' is not running on ${port} / ${addr}: \`dev up ${options.projectName}\` to start it.`);
        }
    }
    catch (err) {
        throw new Error(`DevServer check for '${options.projectName}' on ${port} / ${addr} failed (${err})`);
    }
}
// eslint-disable-next-line @typescript-eslint/naming-convention
export function TEST_testResetCheckPort() {
    checkPort = getCheckPortHelper();
}
function getCheckPortHelper() {
    return fallbackCheckPort;
}
function fallbackCheckPort(address, port, timeout) {
    const result = spawnSync('nc', ['-z', '-w', '1', address, port.toString()], {
        timeout,
        stdio: 'ignore',
    });
    return result.status === 0;
}
